sumoslack/api.py

```diff
@@ -358,47 +359,73 @@ def _billing_info(self, user_id):
     # Done to solve issue "Item size has exceeded the maximum allowed size"
     def save_state(self, cursor, data):
-        channel_ids = []
+        # Get frequent channels current page
+        frequent_channel_page_number = self.kvstore.get("frequent_channel_page_number")
+        frequent_channel_page_number = 1 if frequent_channel_page_number is None else frequent_channel_page_number
+
+        frequent_channels = self.kvstore.get(self.get_key() + self.frequent + str(frequent_channel_page_number))
+        frequent_channels = [] if frequent_channels is None or frequent_channels["ids"] is None \
+            else frequent_channels["ids"]
+
+        # Get in-frequent channels current page
+        in_frequent_channel_page_number = self.kvstore.get("in_frequent_channel_page_number")
+        in_frequent_channel_page_number = 1 if in_frequent_channel_page_number is None \
+            else in_frequent_channel_page_number
+
+        infrequent_channels = self.kvstore.get(self.get_key() + self.in_frequent + str(in_frequent_channel_page_number))
+        infrequent_channels = [] if infrequent_channels is None or infrequent_channels["ids"] is None \
+            else infrequent_channels["ids"]
 
+        # Update the frequent and infrequent list as per threshold provided by user
         if data is not None:
             for channel in data:
-                channel_ids.append(channel["channel_id"] + "#" + channel["channel_name"])
-
-        # Set the channels page number count after dividing channels into group of 50.
-        if self.kvstore.get("Channels_Page_Number") is None:
-            number = 1
-        else:
-            number = self.kvstore.get("Channels_Page_Number") + 1
+                channel_id = channel["channel_id"]
+                channel_name = channel["channel_name"]
+                messages_details = self.kvstore.get(channel_id)
+                if self.enable_infrequent_channels \
+                        and messages_details is not None \
+                        and "fetch_oldest" in messages_details \
+                        and get_current_timestamp() - messages_details.get("fetch_oldest") > \
+                        self.infrequent_channel_threshold:
+                    infrequent_channels.append(channel_id + "#" + channel_name)
+                else:
+                    frequent_channels.append(channel_id + "#" + channel_name)
 
-        ids = self.batchsize_chunking(channel_ids, 50)
-        for channels in ids:
-            obj = {"ids": channels, "last_fetched": get_current_timestamp(), "cursor": cursor}
-            self.kvstore.set(self.get_key() + str(number), obj)
-            self.kvstore.set("Channels_Page_Number", number)
-            number = number + 1
+        # segregate list into chunks of User provided chunks and save them in database.
+        self.put_channels_data(frequent_channels, frequent_channel_page_number, self.frequent
```

Comment:
frequent_channels may contain old + new channels, is there a possibility that it may contain duplicates since it's a list.

Reply:
Frequent channel keys will be deleted before each refresh. Thus, everytime we fetch the channel list from slack, new data will be created.